crates/vm2/src/callframe.rs

```diff
@@ -133,17 +135,29 @@ impl<T, W> Callframe<T, W> {
         })
     }
 
+    /// Gets a raw inferred program counter. This value can be garbage if the frame is on an invalid instruction or free panic.
+    #[allow(clippy::cast_possible_wrap)] // false positive: `Instruction` isn't that large
+    pub(crate) fn get_raw_pc(&self) -> isize {
+        // We cannot use `<*const _>::offset_from` because `self.pc` isn't guaranteed to be allocated within `self.program`
+        // (invalid instructions and free panics aren't).
+        let offset_in_bytes =
+            self.pc as isize - ptr::from_ref(self.program.instruction(0).unwrap()) as isize;
+        offset_in_bytes / mem::size_of::<Instruction<T, W>>() as isize
+    }
+
+    // FIXME: can overflow / underflow on invalid instruction / free panic
+    #[allow(clippy::cast_sign_loss, clippy::cast_possible_truncation)]
     pub(crate) fn get_pc_as_u16(&self) -> u16 {
-        unsafe { self.pc.offset_from(self.program.instruction(0).unwrap()) as u16 }
+        self.get_raw_pc() as u16
     }
 
     /// Sets the next instruction to execute to the instruction at the given index.
     /// If the index is out of bounds, the invalid instruction is used.
-    pub fn set_pc_from_u16(&mut self, index: u16) {
+    pub(crate) fn set_pc_from_u16(&mut self, index: u16) {
         self.pc = self
             .program
             .instruction(index)
-            .unwrap_or(invalid_instruction())
+            .unwrap_or_else(invalid_instruction);
```

Comment:
I've initially [removed](https://github.com/matter-labs/vm2/pull/60/commits/9f72c5f95c0c751d56ca659c1bc8e66bdf381b39) `invalid_instruction` in favor of using an instruction inside a `Program` (it's already present as the last instruction in shorter programs); it also allows not to have a dummy `World` implementation for `Infallible`. However, this led to 5–10% regression on some iai benches.

Reply:
You've highlighted an unsoundness bug that I introduced when I put all the per-instruction logic inside the instructions. The Invalid instruction code assumes that the handler isn't called but this is no longer true.

By putting an empty type as W of INVALID_INSTRUCTION, you actually introduce more unsoundness. The handler of `Instruction<(), Infallible>`, is a function that is impossible to call, as its World argument cannot be constructed. But then you cast it an instruction that can be called.

Reply:
We could store an invalid instruction in the VM like done with panic.

The perhaps nicer alternative is always calling panic and invalid instruction directly, skipping the dispatch loop.

Reply:
> The perhaps nicer alternative is always calling panic and invalid instruction directly, skipping the dispatch loop.

I think this can work for panic; AFAICT, setting `pc` to point to panic is always invoked in an instruction handler and is followed by the handler termination. But in the case of invalid instructions, could you explain how it would work e.g., in this place (i.e., outside an instruction handler)?

Reply:
Ok, I'll fix the unsoundness separately then.